hydrolib/core/basemodel.py

```diff
@@ -778,31 +852,77 @@ def _get_identifier(self, data: dict) -> Optional[str]:
             return filepath.name
         return None
 
-    @property
-    def _relative_mode(self) -> ResolveRelativeMode:
-        """Gets the ResolveRelativeMode of this FileModel.
 
-        Returns:
-            ResolveRelativeMode: The ResolveRelativeMode of this FileModel
-        """
-        return ResolveRelativeMode.ToParent
+class DiskOnlyFileModel(FileModel):
+    """DiskOnlyFileModel provides a stub implementation for file based
+    models which are not explicitly implemented within hydrolib.core.
+
+    It implements the FileModel with a void parser and serializer, and a
+    save method which copies the file associated with the FileModel
+    to a new location if it exists.
+
+    We further explicitly assume that when the filepath is None, no
+    file will be written.
+
+    Actual file model implementations *should not* inherit from the
+    DiskOnlyFileModel and instead inherit directly from FileModel.
+    """
+
+    _source_file_path: Optional[Path] = PrivateAttr(default=None)
+
+    def _post_init_load(self) -> None:
+        # After initialisation we retrieve the _resolved_filepath
+        # this should correspond with the actual absolute path of the
+        # underlying file. Only after saving this path will be updated.
+        super()._post_init_load()
+        self._source_file_path = self._resolved_filepath
+
+    def _load(self, filepath: Path) -> Dict:
+        # We de not load any additional data, as such we return an empty dict.
+        return dict()
+
+    def _save(self) -> None:
+        # The target_file_path contains the new path to write to, while the
+        # _source_file_path contains the original data. If these are not the
+        # same we copy the file and update the underlying source path.
+        target_file_path = self._resolved_filepath
+        if self._can_copy_to(target_file_path):
+            target_file_path.parent.mkdir(parents=True, exist_ok=True)  # type: ignore[arg-type]
+            shutil.copy(self._source_file_path, target_file_path)  # type: ignore[arg-type]
+        self._source_file_path = target_file_path
```

Comment:
Is it correct to set the _source_file_path even if _can_copy_to returned false?

Reply:
Yes this is intentionally done so. Once we save, we expect the new source path to be equal to the saved location. In my opinion this provides the most consistent behavior. Imagine the following use case:

1. A user creates a new model (without the `DiskOnlyFileModel` backed data).
2. The user saves the model
3. The user creates the files mentioned in the `DiskOnlyFileModel` objects  at the newly saved location
4. The user saves the model in a different location

I would expect the data to then be saved at the new location.

Alternatively, imagine the following:

1. A user creates a new model (without the `DiskOnlyFileModel` backed data).
2. The user saves the model
3. The user creates the files mentioned in the `DiskOnlyFileModel` objects  at the previous saved location
4. The user saves the model in a different location

I would now expect no files to be copied, as there are no files at the new saved location.

This would only work if you do always update the `self._source_file_path`  after saving.